File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        # let's also have a parallel stack to hold
        # the minimum value so far
        self.minStack = [] 

    def push(self, val: int) -> None:
        self.stack.append(val)
        # calculate the new minimum
        # is it something that was already there or this new value.
        # Also note, if it was already there, then popping this new value
        # should not have affected that one also
        minSoFar = val if self.minStack == [] else min(val, self.minStack[-1])
        self.minStack.append(minSoFar)

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.js
// to create a mean stack, we will make each value a composite one,
// i.e. [val1, minSoFar1], [val2, minSoFar2] and so on
var MinStack = function() {
    this.st = []
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    // 1st el? then it's also MinSoFar
    if(!this.st.length) this.st.push([val, val]);
    else {
        let lastVal = this.st.at(-1);
        this.st.push([val, Math.min(val, lastVal[1])]);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    return this.st.pop()[0];
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.st.at(-1)[0];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.st.at(-1)[1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */